Login to Github before entering repository name

chia-yh opened the pull request:
### Summary:

Attempt to resolve #154 

WATcher allows users to change repositories, i.e., the repository is not fixed like in CATcher. Having users login first, then being able to choose the repository they wish to see would allow for a smoother UX, e.g.

> **Login -> Sign in to Github -> Confirm Login Account -> Enter Repostory URL -> Issues Dashboard**

_Currently displays a minimal login card that redirects users to sign in to Github. Future enhancements to the UI would be desirable._

#### Type of change:

- ✨ New Feature/ Enhancement

### Changes Made:

- Add `login.component`, which extracts the `AuthService::startOAuthProcess()` call from `session-selection.component.ts` that allows users to login to Github
- Add some tests for `login.component` in `login.component.spec.test`
- Extract user data setup from `ConfirmLoginComponent::completeLoginProcess()` to `AuthService::setRepo()`, leaving the creation of the user model, after which authentication state is changed to authenticated
- Move `handleAuthSuccess()` from `ConfirmLoginComponent` to `AuthService`, renamed to `handleSetRepoSuccess()`
- Modify `ConfirmLoginComponent` tests as necessary
- Add `repoSetSource`, `repoSetState`, `isRepoSet()` to `PhaseService` to determine whether `app-session-selection` is displayed
- Change `auth.component.html` to show `LoginComponent` instead of `SessionSelectionComponent` by default
- Change `auth.component.html` to prevent current session from being shown on the login card for confirming login account, as no session has been set at that point of the process
-  `initializeRepoNameInTitle()` method call in `HeaderComponent` now subscribes to `repoSetState` instead of `currentAuthState`, and checks `isAuthenticated() && isRepoSet()` before calling the method, `html` uses `isRepoSet()` to display buttons requiring a current repository after one has been set

### Screenshots:

![login_before_select_session](https://github.com/CATcher-org/WATcher/assets/105497890/dc677ae0-32ef-4d77-8cf5-6409bf62064c)

### Proposed Commit Message:

```
Change auth process to login before selecting repo

Users select a repository before login.

WATcher does not fix a single repository, and allows users to change the
selected repository.

Having users login to Github prior to selecting a repository would
facilitate a smoother UX.

Let's allow users to login before choosing a repository.
```

<details><summary>
<h3>Checklist:</h3>
</summary>

- [x] I have tested my changes thoroughly.
- [x] I have created tests for any new code files created in this PR or provided a link to a issue/PR that addresses this.
- [ ] I have added or modified code comments to improve code readability where necessary.
- [ ] I have updated the project's documentation as necessary.

</details>

gycgabriel reviewed (commented):
Tested, works good 👍 
Will check code next.

gycgabriel commented on tests/app/auth/confirm-login/confirm-login.component.spec.ts:
```diff
@@ -64,19 +52,14 @@ describe('ConfirmLoginComponent', () => {
+  it('should complete login process', () => {
     userService.createUserModel.and.returnValue(of(mockUser));
-    phaseService.currentRepo = getRepoWithValidName();
     githubService.isRepositoryPresent.and.returnValue(of(true));
```
No repository relevant here

gycgabriel commented on src/app/auth/confirm-login/confirm-login.component.ts:
```diff
@@ -23,12 +16,9 @@ export class ConfirmLoginComponent implements OnInit {
-    private githubEventService: GithubEventService,
     private logger: LoggingService,
-    private router: Router,
     public githubService: GithubService
```
Do not need githubService

gycgabriel commented on tests/app/auth/confirm-login/confirm-login.component.spec.ts:
```diff
@@ -21,35 +18,26 @@ const getRepoWithValidName = () => new Repo('mock', 'repo');
   let fixture: ComponentFixture<ConfirmLoginComponent>;
 
   beforeEach(async(() => {
     authService = jasmine.createSpyObj<AuthService>('AuthService', ['changeAuthState', 'setTitleWithPhaseDetail', 'startOAuthProcess']);
```
Let's remove methods not spied on

gycgabriel commented on tests/app/auth/confirm-login/confirm-login.component.spec.ts:
```diff
@@ -21,35 +18,26 @@ const getRepoWithValidName = () => new Repo('mock', 'repo');
-    phaseService = jasmine.createSpyObj<PhaseService>('PhaseService', ['initializeCurrentRepository', 'currentRepo']);
     logger = jasmine.createSpyObj<LoggingService>('LoggingService', ['info']);
     userService = jasmine.createSpyObj<UserService>('UserService', ['createUserModel']);
     githubService = jasmine.createSpyObj<GithubService>('GithubService', ['isRepositoryPresent']);
```
Let's remove

gycgabriel commented on tests/app/auth/login/login.component.spec.ts:
```diff
@@ -0,0 +1,42 @@
+  let fixture: ComponentFixture<LoginComponent>;
+
+  beforeEach(async(() => {
+    authService = jasmine.createSpyObj<AuthService>('AuthService', ['changeAuthState', 'setTitleWithPhaseDetail', 'startOAuthProcess']);
```
Only include methods to spy on

gycgabriel commented on src/app/auth/confirm-login/confirm-login.component.ts:
```diff
@@ -44,49 +34,22 @@ export class ConfirmLoginComponent implements OnInit {
     this.authService.changeAuthState(AuthState.AwaitingAuthentication);
-    this.phaseService.initializeCurrentRepository();
-    this.logger.info(`ConfirmLoginComponent: Current repo is ${this.phaseService.currentRepo}`);
     this.userService
```
Let's add a log message, start login process at this component

gycgabriel commented on src/app/auth/login/login.component.ts:
```diff
@@ -0,0 +1,30 @@
+    } catch (error) {
+      this.errorHandlingService.handleError(error);
+      this.authService.changeAuthState(AuthState.NotAuthenticated);
+    }
```
Let's add logger.info similar to confirm-login.component.ts if there is error

gycgabriel commented on tests/app/auth/login/login.component.spec.ts:
```diff
@@ -0,0 +1,42 @@
+  it('should complete login process', () => {
+    component.startLoginProcess();
+
+    expect(authService.startOAuthProcess).toHaveBeenCalled();
```
Also expect logger.info

gycgabriel commented on src/app/auth/auth.component.html:
```diff
@@ -1,13 +1,17 @@
     <mat-card-header style="margin-bottom: 20px">
       <mat-card-title class="login-title"> Confirm Login Account </mat-card-title>
-      <mat-card-subtitle> Session on {{ currentSessionOrg }} </mat-card-subtitle>
+      <!-- <mat-card-subtitle> Session on {{ currentSessionOrg }} </mat-card-subtitle> -->
```
Can delete this

gycgabriel commented on src/app/auth/session-selection/session-selection.component.ts:
```diff
@@ -82,13 +80,7 @@ export class SessionSelectionComponent implements OnInit {
-    } catch (error) {
-      this.errorHandlingService.handleError(error);
-      this.authService.changeAuthState(AuthState.NotAuthenticated);
-      this.isSettingUpSession = false;
```
Should keep this

gycgabriel reviewed (approved):
LGTM

gycgabriel commented:
Let's have at least two reviews before merging this PR.

chia-yh commented:
While working on https://github.com/CATcher-org/WATcher/issues/154#issuecomment-1637021639 , it occurred to me that I could make some changes to the current PR that would result in what I believe to be an improvement to the current implementation. I've made the changes in this [branch](https://github.com/chia-yh/WATcher/tree/login-before-select-session-alt):

- `AuthComponent`: removed `SessionSetupStateSubscription` and associated variables/methods, replaced with `isRepoSet()`, which instead simply checks whether a repo has been set, which determines whether `app-session-selection` is displayed by `AuthComponent`
- `AuthService`: replaced `sessionSetupStateSource`, `sessionSetupState`, etc. with `repoSetSource`, `repoSetState`, `isRepoSet()`
- `HeaderComponent`: `initializeRepoNameInTitle()` method call now subscribes to `repoSetState` instead of `currentAuthState`, and checks `isAuthenticated() && isRepoSet()` before calling the method, `html` uses `isRepoSet()` to display buttons requiring a current repository after one has been set
- Modified `ConfirmLoginComponent` tests as necessary

These changes remove the code for setting up a separate subscription in `AuthComponent`, and instead checks if a repo has been set to determine when `SessionSelectionComponent` should be brought up. The initialization of the repository name in the title is also now tied to an observable that represents whether a repository has been set instead of the state of authentication. This allows the state to be set to `AuthState.Authenticated` right after the user has been authenticated, instead of after the user has been authenticated **and** the repository has been set, which I feel represents the current state of WATcher more accurately. In terms of visible changes to the user, as the visual changes in `HeaderComponent` is tied to `isAuthenticated()`, the session selection screen after the user is confirmed now looks like this:
![image](https://github.com/CATcher-org/WATcher/assets/105497890/0615f4c5-109d-43c7-917a-8ad3ad66468e)

instead of like this (current state of the PR):
![image](https://github.com/CATcher-org/WATcher/assets/105497890/8f615f62-f07b-49f3-929a-72c65b39c140)

The user is able to logout at this point, but is unable to jump to a repository or sync with Github as there is no repository set, and "(Issue Dashboard)" is displayed, showing the current phase WATcher is in.

If the changes in the branch seem appropriate, I can merge the commits from the branch onto this PR's branch.

gycgabriel commented:
@chia-yh I will need to look at the code changes to decide, you can make a PR from that branch to this and link it for me to check, or change this branch directly.

chia-yh commented:
> @chia-yh I will need to look at the code changes to decide, you can make a PR from that branch to this and link it for me to check, or change this branch directly.

I've merged the changes onto this branch

gycgabriel commented on src/app/core/services/auth.service.ts:
```diff
@@ -133,6 +143,43 @@ export class AuthService {
+    this.router.navigateByUrl(Phase.issuesViewer);
+  }
+
+  /**
+   * Setup user data after authentication.
+   */
+  setupUserData(): void {
```
Not set up User Data, set Repostory

gycgabriel reviewed (changes requested):
We are decoupling setting repository from auth, so it is better to move setting repository stuff to phase.service.ts

chia-yh commented on src/app/auth/auth.component.ts:
```diff
@@ -134,6 +134,10 @@ export class AuthComponent implements OnInit, OnDestroy {
 
+  isRepoSet(): boolean {
+    return !(this.phaseService.currentRepo === undefined);
+  }
```
I've renamed the methods as suggested.

I also moved `isRepoSet()` in `AuthService` into `PhaseService`, however, I've left the `isRepoSet()` in `AuthComponent` as is as I'm reluctant to make `phaseService` in `AuthComponent` public so that its accessible in `auth.component.html`. I've changed  it to make use of `isRepoSet()` in `PhaseService` though.

As for `SetRepo` and `handleSetRepoSuccess`, I've left them in `AuthService` for now as aside from calling `PhaseService::initializeCurrentRepository()`, they mostly concern  cleaning up before routing to `Phase.issuesViewer`. I'm not too sure how I should break it down further between `AuthService` and `PhaseService`.

Eclipse-Dominator reviewed (changes requested):
The select repo component is not hidden in the page.

Scrolling down on the logging page will reveal the component

Eclipse-Dominator commented on src/app/core/services/auth.service.ts:
```diff
@@ -133,6 +135,37 @@ export class AuthService {
+    this.githubService.isRepositoryPresent(currentRepo.owner, currentRepo.name)
+      .pipe(
+        mergeMap((isValidRepository) => {
+          if (!isValidRepository) {
+            return new Observable();
+          }
+          return this.githubEventService.setLatestChangeEvent();
+        })
+      )
+      .subscribe(
+        () => {
+          this.handleSetRepoSuccess();
+        }
```
```suggestion
          if (!isValidRepository) {
            throw new Error("Repo is invalid!");
          }
          return this.githubEventService.setLatestChangeEvent();
        })
      )
      .subscribe(
        () => this.handleSetRepoSuccess(),
        (error) => {
          // some kind of error handling
          this.logger.error(`AuthService: Error while setting repository: ${error}`);
        }
```

Might be more reasonable to rephrase this since this subscription will be pointless as it will always run `this.handleSetRepoSuccess()` regardless of whether or not the repo is valid.

chia-yh replied:
I played around with the implementation of `setRepo` and decided to shift the check for `isValidRepository` into `PhaseService::initializeCurrentRepository`, so that no repo is set if the repo provided is invalid. `handleSetRepoSuccess` is also now only called if a repo is set in `initializeCurrentRepository`. I've also made some changes in `session-selection.component.ts` so users can try to set a valid repository if they previously provided an invalid one.

chia-yh replied:
> There seems to be 'other forces' at work here as the application is still navigating to issueViewer before `handleSetRepoSuccess()` is not called. (not sure why this is the case)

Also, if you happen to mean the phase descriptor in the header component changing, then I think its because it displays the current phase once the user is authenticated, which by default is the issue dashboard. I'm not sure if i should change it.

Eclipse-Dominator replied:
Ah I see, I think it's fine for now.

Eclipse-Dominator reviewed (changes requested):
Right now, invalid repository like ("    ") will always bring the user to the issueViewer phase. 
This will then lead to a lot of error prompts.

![image](https://github.com/CATcher-org/WATcher/assets/4567895/b9509fe5-afc3-4ffa-9643-5bfbe2dc87e7)
(Empty repo name)

---
There seems to be 'other forces' at work here as the application is still navigating to issueViewer before `handleSetRepoSuccess()` is not called. (not sure why this is the case)

chia-yh commented on src/app/core/services/auth.service.ts:
```diff
@@ -133,6 +135,37 @@ export class AuthService {
+          this.handleSetRepoSuccess();
+        }
+      );
+    this.handleSetRepoSuccess();
```
I see, I was trying to preserve the structure based on the previous `completeLoginProcess()`, but I'll change it so there isn't a call regardless

Eclipse-Dominator commented on src/app/auth/login/login.component.ts:
```diff
@@ -0,0 +1,31 @@
+    private logger: LoggingService
+  ) {}
+
+  ngOnInit() {}
```
```suggestion
```

`ngOnInit()` is not used. Let's remove `implements OnInit`!

Eclipse-Dominator reviewed (dismissed):
LGTM!

chia-yh commented:
@gycgabriel could I check if there are any further comments with respect to changes that need to be made to this PR?

gycgabriel reviewed (approved):
LGTM, my comments have been addressed.

gycgabriel reviewed (dismissed):
LGTM

gycgabriel commented on src/app/core/services/auth.service.ts:
```diff
@@ -69,7 +71,6 @@ export class AuthService {
     this.githubService.reset();
     this.userService.reset();
     this.issueService.reset(true);
-    this.labelService.reset();
```
What's the reason for removing this label service?

chia-yh replied:
I'm not sure why, but at the time I was under the impression that it wasn't necessary, which I realize might not be the case.

I've made a commit reverting the changes, sorry about that. ">_>

gycgabriel reviewed (approved):
LGTM
